ofrak_core/ofrak/core/esp.py

```diff
@@ -0,0 +1,779 @@
+    +--------+------------------------------------------------------------------+
+    | 4-5    | Chip ID (which ESP device is this image for)                     |
+    +--------+------------------------------------------------------------------+
+    | 6      | Min chip rev supported by image (deprecated, use following field)|
+    +--------+------------------------------------------------------------------+
+    | 7-8    | Min chip rev supported by the image (in format: major*100+minor) |
+    +--------+------------------------------------------------------------------+
+    | 9-10   | Max chip rev supported by the image (in format: major*100+minor) |
+    +--------+------------------------------------------------------------------+
+    | 11-14  | Reserved bytes in additional header space, currently unused      |
+    +--------+------------------------------------------------------------------+
+    | 15     | Hash appended (If 1, SHA256 digest appended after the checksum)  |
+    +--------+------------------------------------------------------------------+
+    Note: the spi drive settings are auto parsed into clk_drv, q_drv, d_drv,
+    cs_drv, hd_drv, and wp_drv.
+
+## Segment:
+    +---------+-----------------+
+    | Byte    | Description     |
+    +=========+=================+
+    | 0-3     | Memory offset   |
+    +---------+-----------------+
+    | 4-7     | Segment size    |
+    +---------+-----------------+
+    | 8...n   | Data            |
+    +---------+-----------------+
+
+## Footer:
+The file is padded with zeros until its size is one byte less than a multiple of
+16 bytes. A last byte (thus making the file size a multiple of 16) is the
+checksum of the data of all segments. The checksum is defined as the xor-sum of
+all bytes and the byte ``0xEF``.
+
+### Not in ESP8266:
+
+    If ``hash appended`` in the extended file header is ``0x01``, a SHA256
+    digest “simple hash” (of the entire image) is appended after the checksum.
+    This digest is separate to secure boot and only used for detecting
+    corruption. The SPI flash info cannot be changed during flashing if hash is
+    appended after the image.
+
+    If secure boot is enabled, a signature is also appended (and the simple hash
+    is included in the signed data). This image signature is `Secure Boot V1
+    <https://docs.espressif.com/projects/esp-idf/en/latest/esp32/security/secure-boot-v1.html#image-signing-algorithm>`_
+    and `Secure Boot V2
+    <https://docs.espressif.com/projects/esp-idf/en/latest/esp32/security/secure-boot-v2.html#signature-block-format>`_
+    specific.
+
+## Analyzing a Binary Image
+To analyze a binary image and get a complete summary of its headers and
+segments, use the :ref:`image_info <image-info>` command with the ``--version
+2`` option.
+"""
+
+LOGGER = logging.getLogger(__name__)
+
+ESP_BINARY_MAGIC = 0xE9
+ESP8266V2_BINARY_MAGIC = 0xEA  # In the esptool.py code, but haven't seen it in the documentation
```

Comment:
I'm not sure about this, 0xE9 and 0xEA will match a lot of arm instructions, we might just have to tag manually if there isn't a larger magic we can find.

Reply:
I was concerned about that as well. I could implement a more heuristic based approach. Byte 2 (zero index) is the flash mode which is always 0,1, 2, or 3 and the lower half of byte three is always 0, 1, 2, or 0xf. If you think that would narrow it down enough I can implement that. If we go the manual tagging route, do I just remove the identify function?

Reply:
The more constraints, the better.

If manual tagging turns out to be the only way to avoid a very high false positive ID rate, then the Identifier component as a whole can be removed.